test/dist_autograd_test.py

```diff
@@ -231,20 +405,29 @@ def test_worker_ids_recorded(self):
             for dst_rank in dst_ranks:
                 ret = rpc.rpc_sync("worker{}".format(dst_rank), torch.add, args=(t1, t2))
                 rpc.rpc_sync(
-                    "worker{}".format(dst_rank), _set_rpc_done, args=(context_id,)
+                    "worker{}".format(dst_rank), _set_rpc_done, args=(context_id, 1)
                 )
             # no worker ids should be recorded.
+            # no send and recv functions are attached.
             ctx = dist_autograd._current_context()
             worker_ids = ctx._known_worker_ids()
             self.assertEqual(len(worker_ids), 0)
+            send_functions = ctx._send_functions()
+            self.assertEqual(len(send_functions), 0)
+            recv_functions = ctx._recv_functions()
+            self.assertEqual(len(recv_functions), 0)
+            # prev context id is not passed over as tensors do not require grads
+            with self.assertRaises(RuntimeError):
+                ctx = dist_autograd._retrieve_context(ctx_ids[1])
+
```

Comment:
I think we need a `dist.barrier()` here, since this test is modified to test for state on a process that is set by an RPC from another process (previously this was just testing local state). Without `dist.barrier()`, for example, worker 0 could run the 2nd portion of this test (where the tensors do require gradients), and create the context on worker 1. Then worker 1 can run in the 1st portion of the test, and the assert would fail.

Reply:
ah, my bad, forgot to call "self._check_rpc_done(1)" before calling _retrieve_context(ctx_ids[1])